### PR TITLE
[AIP-158] Response's first field should be repeated.

### PR DESCRIPTION
**Pull request by Dos-Ph**
Issue #490 

For all messages that:
- end in Response
- Contain a next_page_token field.

Throw a linter warning if: 
The first field in the message (by both position and field number) is not repeated

### CODE REVIEW

File: docs/rules/0158/response-repeated-first-field.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+# Paginated methods: Page token field
+
+This rule enforces that all `List` and `Search` methods have a repeatable field
+as a first field in the response message, as mandated in [AIP-158][].
+
+## Details
+
+This rule looks at any message matching `List*Response` or `Search*Response`
+that has `next_page_token` field and complains if the first field (by both
```
Comment (jgeewax) [inline docs/rules/0158/response-repeated-first-field.md]: Why does it matter if they have a `next_page_token` field? For example, shouldn't we warn on the following:

```proto
message ListBooks {
  Book book = 1;
}
```

and say that the correct version is

```proto
message ListBooks {
  repeated Book books = 1;
}
```

?

File: docs/rules/0158/response-repeated-first-field.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+## Examples
+
+**Incorrect** code for this rule:
+
+```proto
+// Incorrect
+message ListStudentProfilesResponse {
+    Profile studentProfiles = 1; // Field should be repeated.
+    string next_page_token = 2;
```
Comment (jgeewax) [inline docs/rules/0158/response-repeated-first-field.md]:
```suggestion
  Profile student_profiles = 1; // Field should be repeated.
  string next_page_token = 2;
```

File: docs/rules/0158/response-repeated-first-field.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+```
+
+**Correct** code for this rule:
+
+```proto
+// Correct.
+message ListStudentProfilesResponse {
+  repeated Profile studentProfiles = 1;
```
Comment (jgeewax) [inline docs/rules/0158/response-repeated-first-field.md]:
```suggestion
  repeated Profile student_profiles = 1;
```

File: docs/rules/0158/response-repeated-first-field.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+If you need to violate this rule, use a leading comment above the message or
+above the field. Remember to also include an [aip.dev/not-precedent][] comment
+explaining why.
+
+```proto
+// (-- api-linter: core::0158::response-repeated-first-field
+//     aip.dev/not-precedent: We need to do this because reasons. --)
+message ListStudentProfilesResponse {
+    Profile studentProfiles = 1; // Field should be repeated.
+    string next_page_token = 2;
```
Comment (jgeewax) [inline docs/rules/0158/response-repeated-first-field.md]:
```suggestion
  Profile student_profiles = 1; // Field should be repeated.
  string next_page_token = 2;
```

File: rules/aip0158/response_repeated_first_field.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/jhump/protoreflect/desc"
+)
+
+var responseRepeatedFirstField = &lint.MessageRule{
+	Name:   lint.NewRuleName(158, "response-repeated-first-field"),
+	OnlyIf: isPaginatedResponseMessage,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		// Rule check: for all messages that end in Response and contain a next_page_token field.
+		// Throw a linter warning if, the first field in the message (by both position and field number) is not repeated
+		nextPageToken := m.FindFieldByName("next_page_token")
```
Comment (jgeewax) [inline rules/aip0158/response_repeated_first_field.go]: See comment from above. Maybe there's a good reason to require a check for the `next_page_token` field?

File: rules/aip0158/response_repeated_first_field.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		if nextPageToken == nil {
+			return nil
+		}
```
Comment (jgeewax) [inline rules/aip0158/response_repeated_first_field.go]: We don't need to hold onto the result of the value, right? Can we in-line this? Or am I being a bad gopher?

```suggestion
		if m.FindFieldByName("next_page_token") == nil {
			return nil
		}
```

File: rules/aip0158/response_repeated_first_field.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/googleapis/api-linter/lint"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var responseRepeatedFirstField = &lint.MessageRule{
+	Name:   lint.NewRuleName(158, "response-repeated-first-field"),
+	OnlyIf: isPaginatedResponseMessage,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		// Rule check: for all messages that end in Response and contain a next_page_token field.
+		// Throw a linter warning if, the first field in the message (by both position and field number) is not repeated
```
Comment (jgeewax) [inline rules/aip0158/response_repeated_first_field.go]:
```suggestion
		// Throw a linter warning if the first field in the message (by both position and field number) is not repeated
```

File: rules/aip0158/response_repeated_first_field.go (hunk 5 above)
Comment (Dos-Ph) [reply rules/aip0158/response_repeated_first_field.go]: @lukesneeringer should know better. summon!

File: docs/rules/0158/response-repeated-first-field.md (hunk 2 above)
Comment (Dos-Ph) [reply docs/rules/0158/response-repeated-first-field.md]: Oops. Shame on me. T___T

File: docs/rules/0158/response-repeated-first-field.md (hunk 1 above)
Comment (lukesneeringer) [reply docs/rules/0158/response-repeated-first-field.md]: Maybe make it a disjunction? I definitely want to match any response message that has `string next_page_token` even if it does not have a message name I recognize. It seems reasonable that I also want to match any ListFooResponse or SearchFooResponse

File: rules/aip0158/response_repeated_first_field.go (hunk 5 above)
Comment (lukesneeringer) [reply rules/aip0158/response_repeated_first_field.go]: I was planning on checking for `string next_page_token` _instead of_ checking for ListFooResponse or SearchFooResponse as message names.

I would be fine with a disjunction. I would also be fine with just `string next_page_token`, since this is the pagination AIP after all.

Slight preference for just `string next_page_token`.

File: rules/aip0158/response_repeated_first_field.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	Name:   lint.NewRuleName(158, "response-repeated-first-field"),
+	OnlyIf: isPaginatedResponseMessage,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		// Rule check: for all messages that end in Response and contain a next_page_token field.
+		// Throw a linter warning if the first field in the message (by both position and field number) is not repeated
+		if m.FindFieldByName("next_page_token") == nil {
+			return nil
+		}
+
+		if !m.FindFieldByNumber(1).IsRepeated() {
```
Comment (mingzhi) [inline rules/aip0158/response_repeated_first_field.go]: Can you combine checking first field by number and by position into one? Like:
```golang
// check if the message is empty
if len(m.GetFields()) == 0 {
 ...
}

firstField := m.FindFieldByNumber(1)
if firstField == nil {
  firstField := m.GetFields()[0]
}

if !firstField.IsRepeated() {
  ...
}

File: rules/aip0158/response_repeated_first_field_test.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+package aip0158
+
+import (
+	"testing"
+
+	"github.com/googleapis/api-linter/rules/internal/testutils"
+)
+
+func TestResponseRepeatedFirstField(t *testing.T) {
```
Comment (mingzhi) [inline rules/aip0158/response_repeated_first_field_test.go]: Can you add a test for empty message? Like:
```protobuf
message ListBooksResponse{}
```

Comment (jgeewax) [reply rules/aip0158/response_repeated_first_field_test.go]: What's the right situation here? I'd argue that a ListBooksResponse should **require** that the first field be repeated. If it's missing, then that seems like a failure to me. Currently we will ignore this and say "OK, no `next_page_token` field, so therefore this is perfectly fine."

File: docs/rules/0158/response-repeated-first-field.md (hunk 1 above)
Comment (Dos-Ph) [reply docs/rules/0158/response-repeated-first-field.md]: Discussed in this PR: https://github.com/googleapis/api-linter/pull/496

File: rules/aip0158/response_repeated_first_field.go (hunk 8 above)
Comment (Dos-Ph) [reply rules/aip0158/response_repeated_first_field.go]: If we combine it, we will miss this case.
```
message Profile {
  string name = 1;
}
message ListStudentProfilesResponse {
  Profile student_profiles = 2;
  repeated string next_page_token = 1;
}
` ``

File: rules/aip0158/response_repeated_first_field.go (hunk 5 above)
Comment (Dos-Ph) [reply rules/aip0158/response_repeated_first_field.go]: Removed the next_page_token check. It is covered by `isPaginatedResponseMessage` function now.
https://github.com/googleapis/api-linter/blob/master/rules/aip0158/aip0158.go#L44-L47

File: rules/aip0158/response_repeated_first_field_test.go (hunk 9 above)
Comment (Dos-Ph) [reply rules/aip0158/response_repeated_first_field_test.go]: I added warning message for an empty response

File: rules/aip0158/response_repeated_first_field.go (hunk 8 above)
Comment (mingzhi) [reply rules/aip0158/response_repeated_first_field.go]: I see, thank you!

File: rules/aip0158/response_repeated_first_field_test.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+			"InvalidType",
+			"",
+			"",
+			testutils.Problems{{}}},
+		{
+			// first field by number should be repeated.
+			"InvalidType",
+			"repeated string next_page_token = 2;",
+			"Book books = 1;",
+			testutils.Problems{{}}},
```
Comment (mingzhi) [inline rules/aip0158/response_repeated_first_field_test.go]: Should we expect a finding/problem here?

Comment (Dos-Ph) [reply rules/aip0158/response_repeated_first_field_test.go]: Yes, we expect a problem here. The problem detail is added in line 68-74.

Comment (mingzhi) [reply rules/aip0158/response_repeated_first_field_test.go]: (Sorry for the late response) I suggest you to remove the `problems` field in the test cases, as it provides no information about the test cases.

Comment (Dos-Ph) [reply rules/aip0158/response_repeated_first_field_test.go]: The `problems` field is used. It is set at line 69.

**Comment by Dos-Ph** [comment]
@lukesneeringer @jgeewax @mingzhi Hi, I wonder if there's anything else that you think I missed on this PR?

File: docs/rules/0158/response-repeated-first-field.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
+---
+rule:
+  aip: 158
+  name: [core, '0158', response-repeated-first-field]
+  summary:
```
Comment (lukesneeringer) [inline docs/rules/0158/response-repeated-first-field.md:5]:
```suggestion
  summary: |
```

File: docs/rules/0158/response-repeated-first-field.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+This rule looks at any message matching `List*Response` or `Search*Response`
+that has `next_page_token` field and complains if the first field (by both
+position and field number) is not repeated.
+
+## Examples
+
+**Incorrect** code for this rule:
+
+```proto
+// Incorrect
```
Comment (lukesneeringer) [inline docs/rules/0158/response-repeated-first-field.md:29]:
```suggestion
// Incorrect.
```

File: docs/rules/0158/response-repeated-first-field.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+message ListBooksResponse {
+    Book books = 1; // Field should be repeated.
+    string next_page_token = 2;
```
Comment (lukesneeringer) [inline docs/rules/0158/response-repeated-first-field.md:32]:
```suggestion
  Book books = 1;  // Field should be repeated.
  string next_page_token = 2;
```

File: rules/aip0158/response_repeated_first_field.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"github.com/googleapis/api-linter/lint"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var responseRepeatedFirstField = &lint.MessageRule{
+	Name:   lint.NewRuleName(158, "response-repeated-first-field"),
+	OnlyIf: isPaginatedResponseMessage,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		if len(m.GetFields()) == 0 {
```
Comment (lukesneeringer) [inline rules/aip0158/response_repeated_first_field.go:26]: Include this in `OnlyIf` (and send nothing).

File: rules/aip0158/response_repeated_first_field.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+			return []lint.Problem{{
+				Message: "Paginated RPCs' response should have at least 1 field.",
+			}}
+		}
+
+		if !m.FindFieldByNumber(1).IsRepeated() {
```
Comment (lukesneeringer) [inline rules/aip0158/response_repeated_first_field.go:32]: This will succeed if field 1 and the first field are _both_ repeated, but not the same field. That's bad too.
A better approach is probably:

```go
if m.FindFieldByNumber(1) != m.GetFields()[0] {
  return []lint.Problem{{
    Message: "For list responses, the first field must be proto field 1.",
    Descriptor: m.GetFields()[0],
  }}
}

if !m.GetFields()[0].IsRepeated() {
  return []lint.Problem{{
    Message: "The first field of paginated RPCs must be repeated.",
    Descriptor: m.GetFields()[0],
  }}
}
```

File: rules/aip0158/response_repeated_first_field_test.go (hunk 10 above)
Comment (lukesneeringer) [reply rules/aip0158/response_repeated_first_field_test.go:46]:
```suggestion
			testutils.Problems{{Message: ""be repeated"}}},
```

File: rules/aip0158/response_repeated_first_field_test.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+		{
+			// first field by position should be repeated.
+			"InvalidType",
+			"Book books = 2;",
+			"repeated string next_page_token = 1;",
+			testutils.Problems{{}}},
```
Comment (lukesneeringer) [inline rules/aip0158/response_repeated_first_field_test.go:52]:
```suggestion
			testutils.Problems{{Message: ""be repeated"}}},
```

**Comment by lukesneeringer** [comment]
In the interest of getting this in, I made the needed fixes. :-)
